pkgs/development/compilers/gcc/common/configure-flags.nix

```diff
@@ -114,6 +114,14 @@ let
     ++ lib.optional (!(crossMingw && crossStageStatic))
       "--with-native-system-header-dir=${lib.getDev stdenv.cc.libc}/include"
 
+    # cross-built gcc needs to be tricked as if it is built with headers
```

Comment:
Can you provide some insight on why the fix is not platform-specific, yet the problem only occurs on powerpc64le?  The problem that started this PR doesn't occur on mips64el, which also has a cross-compiled bootstrap-files bundle.  Why didn't mips64el have the same problem?

```
$ nix log -f pkgs/stdenv/linux/make-bootstrap-tools-cross.nix  mips64el-linux-gnuabi64.bootGCC | fgrep __stack_chk_fail
checking __stack_chk_fail in target C library... no
checking __stack_chk_fail in target C library... no
```

I would like to understand this in case something similar happens in the future.

Reply:
Also wondering why the riscv64 people haven't hit this.

It looks like aarch64 has never used cross-built bootstrap-files; the "seed" for that platform is this commit 5c0a385e1cafb369529bb304ff8600c1f44b3fc3 rather than a cross-build from x86_64.

Reply:
I can't explain why this happens, but I can say that it shows up on more than just powerpc64le. For example, my linked PR https://github.com/NixOS/nixpkgs/issues/113977 shows how to replicate this in a i686 -> x86_64 cross scenario.

Reply:
## A bit of generic background before we dive into mips64el details

Fundamentally stack protector slightly modifies original C code with an equivalent like:

```c
void func() { <code>; }
```

to something like:

```c
void func() {
    long canary = __stack_chk_guard();
    <code>;
    if (canary != __stack_chk_guard()) __stack_check_fail().
}
```

The compiler can either generate special code in place of __underscodres or use it as is.

The trick is where to pick a canary. Ideally it should be a random per-thread value. For example on x86_64 kernel provides that as a fixed offset of TLS page: `__stack_chk_guard == %gs:0x14`, `powerc64le` does `-0x7008(2)` (`r2` is a TOC pointer, not sure who points value there). The takeaway here is that reading a value might not require any symbol relocations: it's a somewhat simple memory read.

The fallback code has no these nice properties: it uses global variable (which everyone could read or modify, main security concern) and possibly calls external function:

```c
extern void *__stack_chk_guard; /* from libssp? */
void func() {
    long canary = __stack_chk_guard;
    <code>;
    if (canary != __stack_chk_guard) __stack_check_fail().
}
```

This means that the ABI assumptions of code compiled by different backend implementations of the same target platform are slightly different:
- global variable `__stack_chk_guard` either has to exist or not
- global function `__stack_check_fail()` either has to exist of not (sometimes targets use module-local call to avoid registrer pressure and PLT on this slow path, i386 and ppc32 does it and has some troubles on musl: https://gcc.gnu.org/legacy-ml/gcc/2020-01/msg00454.html)

Mixing code from both usually causes complains about missing symbols that one backend generates as external symbols (and assumes `libssp` to be used) and another implements as builtins (and avoids `libssp`).

Most of the time throwing `llssp` in (or the equivalent) feels like it works. But it's more of a workaround.

## Let's check what mips64el does in both cases

Could it be that `mips` has no target specific code at all and always falls back to generic `libssp` code?

Building two compilers: cross-compiler (usually works), cross-build (expected to be problematic):

```
$ nix-build -A pkgsCross.mips64el-linux-gnuabin32.stdenv.cc -A pkgsCross.mips64el-linux-gnuabin32.gcc
```

Checking both:

cross-compiler (usually works):
```
$ printf "int main(){}" | ./result/bin/mips64el-unknown-linux-gnuabin32-gcc -S -x c - -o - -fstack-protector-all | fgrep stack
        lw      $2,%got_disp(__stack_chk_guard)($28)
        lw      $2,%got_disp(__stack_chk_guard)($28)
        lw      $2,%call16(__stack_chk_fail)($28)
        .reloc  1f,R_MIPS_JALR,__stack_chk_fail
        .section        .note.GNU-stack,"",@progbits
```

It reads `__stack_chk_guard` global as is and calls `__stack_chk_fail` as is.

cross-built compiler:
```
        lw      $2,%got_disp(__stack_chk_guard)($28)
        lw      $2,%got_disp(__stack_chk_guard)($28)
        lw      $2,%call16(__stack_chk_fail)($28)
        .reloc  1f,R_MIPS_JALR,__stack_chk_fail
```

Both look almost identical. That means mips does not have any special stack protector code (TIL).

[ I don't like that missing gnustack. I suspect I broke header detection again. Will double-check tomorrow. ]

Reply:
> [ I don't like that missing gnustack. I suspect I broke header detection again. Will double-check tomorrow. ]

Oh, that's expected as this was a test on a tree without this PR applied to demonstrate there is no material code generation difference.

Reply:
> Could it be that `mips` has no target specific code at all and always falls back to generic `libssp` code?

> That means mips does not have any special stack protector code (TIL).

Awesome, this explains everything!  Please mark this thread resolved.

Reply:
> Mixing code from both usually causes complains about missing symbols that one backend generates as external symbols (and assumes `libssp` to be used) and another implements as builtins (and avoids `libssp`).

This is something that worries me a lot about our current bootstrap process.  We pass `CC=${bootstrap-files}/bin/gcc` when building `${stdenv.cc}/bin/gcc` using gcc's built-in `stage1`/`stage2` process.  So you could think of `${bootstrap-files}/bin/gcc` as "`stage0`" -- it's what compiles `stage1`-gcc.

However the `lib{mpfr,isl,mpc,gmp}.a` which gets linked into `stage2`-gcc were *not* built by `stage1`-gcc.  They were built by `${bootstrap-files}/bin/gcc`.

So far this has been a mostly-theoretical concern, but I've been keeping a [list of worries](https://github.com/NixOS/nixpkgs/pull/170857#issuecomment-1170558348) it causes.  You've identified an additional one; thanks!

Reply:
Yeah, I agree `nixpkgs`'s linux bootstrap is a bit fragile today. It would certainly not work if we tried to bootstrap glibc from musl bootstrap tools for example (or even from very different glibc), or bootstrap amd64 from i686.

It would be nice to eventually make it work LFS-style where all of stage1 (a bit wider "stage" definition: includes gcc, binutils, glibc and their dependencies) gets built with stage0 alone, and stage2 gets build with stage1 alone. Not sure how many more packages would get pulled in when built like that. Perhaps it's considerably bigger?